source/adapters/cuda/command_buffer.cpp

```diff
@@ -9,48 +9,154 @@
+/// @param[in] SyncPointWaitList List of sync points in \p CommandBuffer
+/// to find the events for.
+/// @param[out] CuNodesList Return parameter for the Cuda Nodes associated with
+/// each sync-point in \p SyncPointWaitList.
+///
+/// @return UR_RESULT_SUCCESS or an error code on failure
+static ur_result_t getNodesFromSyncPoints(
+    const ur_exp_command_buffer_handle_t &CommandBuffer,
+    size_t NumSyncPointsInWaitList,
+    const ur_exp_command_buffer_sync_point_t *SyncPointWaitList,
+    std::vector<CUgraphNode> &CuNodesList) {
+  // Map of ur_exp_command_buffer_sync_point_t to ur_event_handle_t defining
+  // the event associated with each sync-point
+  auto SyncPoints = CommandBuffer->SyncPoints;
+
+  // For each sync-point add associated L0 event to the return list.
+  for (size_t i = 0; i < NumSyncPointsInWaitList; i++) {
+    if (auto NodeHandle = SyncPoints.find(SyncPointWaitList[i]);
+        NodeHandle != SyncPoints.end()) {
+      CuNodesList.push_back(*NodeHandle->second.get());
+    } else {
+      return UR_RESULT_ERROR_INVALID_VALUE;
+    }
+  }
+  return UR_RESULT_SUCCESS;
+}
+
+/// Set parameter for General 1D memory copy.
+/// If the source and/or destination is on the device, SrcPtr and/or DstPtr
+/// must be a pointer to a CUdeviceptr
+static void setCopyParams(const void *SrcPtr, const CUmemorytype_enum SrcType,
+                          void *DstPtr, const CUmemorytype_enum DstType,
+                          size_t Size, CUDA_MEMCPY3D &Params) {
+
+  Params.srcMemoryType = SrcType;
+  Params.srcDevice = SrcType == CU_MEMORYTYPE_DEVICE
+                         ? *static_cast<const CUdeviceptr *>(SrcPtr)
+                         : 0;
+  Params.srcHost = SrcType == CU_MEMORYTYPE_HOST ? SrcPtr : nullptr;
+  Params.srcXInBytes = 0;
+  Params.srcY = 0;
+  Params.srcZ = 0;
+  Params.srcLOD = 0;
+  Params.srcPitch = 0;
+  Params.srcHeight = 0;
+
+  Params.dstMemoryType = DstType;
+  Params.dstDevice =
+      DstType == CU_MEMORYTYPE_DEVICE ? *static_cast<CUdeviceptr *>(DstPtr) : 0;
+  Params.dstHost = DstType == CU_MEMORYTYPE_HOST ? DstPtr : nullptr;
+  Params.dstXInBytes = 0;
+  Params.dstY = 0;
+  Params.dstZ = 0;
+  Params.dstLOD = 0;
+  Params.dstPitch = 0;
+  Params.dstHeight = 0;
+
+  Params.WidthInBytes = Size;
+  Params.Height = 1;
+  Params.Depth = 1;
+}
 
 UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferCreateExp(
     ur_context_handle_t hContext, ur_device_handle_t hDevice,
-    const ur_exp_command_buffer_desc_t *pCommandBufferDesc,
-    ur_exp_command_buffer_handle_t *phCommandBuffer) {
-  (void)hContext;
-  (void)hDevice;
-  (void)pCommandBufferDesc;
-  (void)phCommandBuffer;
-  detail::ur::die("Experimental Command-buffer feature is not "
-                  "implemented for CUDA adapter.");
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    const ur_exp_command_buffer_desc_t *hCommandBufferDesc,
+    ur_exp_command_buffer_handle_t *hCommandBuffer) {
```

Comment:
why are we taking hCommandBuffer as an argument only to immediately assign it a new value?

Reply:
This is `urCreateCommandBufferExp` so the command buffer handle here is what will be returned from the function. I guess it is strongly expected that the handle will be `nullptr` in this case, and I don't think other UR functions check that is actually the case either.

Reply:
So its an out parameter. I think it should be `phCommandBuffer` since its a pointer to the handle.

Reply:
> So its an out parameter. I think it should be `phCommandBuffer` since its a pointer to the handle.

Ah yes good catch, I've fixed that and also `hCommandBufferDesc` which should be `pCommandBufferDesc` as it is in the specification.